### ISSUE TITLE
False positive warnings and validation crash

### ISSUE THREAD

**Issue by mikes-lunarg** [KhronosGroup/Vulkan-ValidationLayers#238]
**Issue by [Plagman](https://github.com/Plagman) (MIGRATED)**
_Tuesday July 10, 2018 at 22:24 GMT_
_Originally opened as https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/2677_

----

If you download this trace:

https://www.dropbox.com/s/ycsy007lli0im5e/witcher3.7z?dl=0

Run it with apitrace.exe with [DXVK binaries](https://github.com/doitsujin/dxvk/releases/download/v0.61/dxvk-0.61.tar.gz) (d3d11.dll and dxgi.gll) dropped into the same directory; observe these sort of errors:

VkPhysicalDeviceFeatures::shaderStorageImageWriteWithoutFormat but is not enabled on the device

I'm told this particular trace also causes a crash when validation is enabled.

**Comment by mikes-lunarg** [comment]
I was finally able to reproduce this issue with the following setup:

- wine 3.11
- apitrace-msvc-latest (https://people.freedesktop.org/~jrfonseca/apitrace/ 2018-07-04)
- DXVK and witcher3.trace from issue descriptoin
- NVIDIA driver 396.24.02 (beta version with VK_EXT_vertex_attribute_divisor required by DXVK)

Resulting in the following unique validation errors:

```
Validation(ERROR): msg_code: -1:  [ UNASSIGNED-CoreValidation-Shader-FeatureNotEnabled ] Object: VK_NULL_HANDLE (Type = 0) | Shader requires extension VK_KHR_shader_draw_parameters but is not enabled on the device
Validation(ERROR): msg_code: -1:  [ UNASSIGNED-CoreValidation-Shader-FeatureNotEnabled ] Object: VK_NULL_HANDLE (Type = 0) | Shader requires VkPhysicalDeviceFeatures::sampleRateShading but is not enabled on the device
Validation(ERROR): msg_code: -1:  [ UNASSIGNED-CoreValidation-Shader-FeatureNotEnabled ] Object: VK_NULL_HANDLE (Type = 0) | Shader requires VkPhysicalDeviceFeatures::shaderClipDistance but is not enabled on the device
Validation(ERROR): msg_code: -1:  [ UNASSIGNED-CoreValidation-Shader-FeatureNotEnabled ] Object: VK_NULL_HANDLE (Type = 0) | Shader requires VkPhysicalDeviceFeatures::shaderCullDistance but is not enabled on the device
Validation(ERROR): msg_code: -1:  [ UNASSIGNED-CoreValidation-Shader-FeatureNotEnabled ] Object: VK_NULL_HANDLE (Type = 0) | Shader requires VkPhysicalDeviceFeatures::shaderStorageImageReadWithoutFormat but is not enabled on the device
Validation(ERROR): msg_code: -1:  [ UNASSIGNED-CoreValidation-Shader-FeatureNotEnabled ] Object: VK_NULL_HANDLE (Type = 0) | Shader requires VkPhysicalDeviceFeatures::shaderStorageImageWriteWithoutFormat but is not enabled on the device
Validation(ERROR): msg_code: 333448904:  [ VUID-VkSubpassDependency-srcAccessMask-00868 ] Object: VK_NULL_HANDLE (Type = 0) | CreateRenderPass: pDependencies[1].srcAccessMask (0x7E0) is not supported by srcStageMask (0x700). The spec valid usage text states 'Any access flag included in srcAccessMask must be supported by one of the pipeline stages in srcStageMask, as specified in the table of supported access types.' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-VkSubpassDependency-srcAccessMask-00868)
Validation(ERROR): msg_code: 333448906:  [ VUID-VkSubpassDependency-dstAccessMask-00869 ] Object: VK_NULL_HANDLE (Type = 0) | CreateRenderPass: pDependencies[0].dstAccessMask (0x7E0) is not supported by dstStageMask (0x700). The spec valid usage text states 'Any access flag included in dstAccessMask must be supported by one of the pipeline stages in dstStageMask, as specified in the table of supported access types.' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-VkSubpassDependency-dstAccessMask-00869)
Validation(ERROR): msg_code: 532679382:  [ VUID-vkCreateDevice-ppEnabledExtensionNames-01387 ] Object: 0x7c89f0c0 (Type = 1) | Missing required extensions for device extension VK_EXT_vertex_attribute_divisor, VK_KHR_get_physical_device_properties2. The spec valid usage text states 'All required extensions for each extension in the VkDeviceCreateInfo::ppEnabledExtensionNames list must also be present in that list.' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-vkCreateDevice-ppEnabledExtensionNames-01387)
Validation(ERROR): msg_code: 532679382:  [ VUID-vkCreateDevice-ppEnabledExtensionNames-01387 ] Object: 0x7cf8b040 (Type = 1) | Missing required extensions for device extension VK_EXT_vertex_attribute_divisor, VK_KHR_get_physical_device_properties2. The spec valid usage text states 'All required extensions for each extension in the VkDeviceCreateInfo::ppEnabledExtensionNames list must also be present in that list.' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-vkCreateDevice-ppEnabledExtensionNames-01387)
```

I'll start digging around with debugger to see why the shader validation code is printing those errors. One final note: I was not able to reproduce the crash. The attached trace exits cleanly with validation enabled.

**Comment by Plagman** [comment]
Thanks for looking into it; I think the original reported mentioned the crash was only on Windows, where you can also reproduce the issue without Wine in the picture.

**Comment by Plagman** [comment]
Sorry, apparently the crash issue is only an issue when running the actual Witcher 3 title with validation enabled through DXVK on Windows, but won't happen with just a trace, unlike these false positive errors; it should be straightforward to set up by dropping the DXVK libraries next to the game executable. I can provide a license for that title if you send me a Steam account name.

**Comment by mikes-lunarg** [comment]
The shader warnings are fixed in https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/247

I was able to reproduce a crash on windows but it appears to be the driver bug previously documented here: https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/2543

I filed a report with my IHV contact but will summarize here as well:

- The game calls vkCreateRenderPass with:
  - pCreateInfo->pSubpasses[0].colorAttachementCount == 8
  - pCreateInfo->pSubpasses[0].pColorAttachements[0-7].attachment == VK_ATTACHEMENT_UNUSED.  
- The game then calls vkCreateGraphicsPipelines with:
  - renderPass == VkRenderPass object created above
  - pCreateInfos[0].pColorBlendState fully populated.

This works fine without layers, but when VK_LAYER_GOOGLE_unique_objects is enabled it has to create a copy of vkCreateGraphicsPipelines pCreateInfos to do handle substitution.  In constructing the copy, the layer sets pColorBlendState to null as it cannot safely follow the the pointer because it could be uninitialized according to the spec:

>pColorBlendState is a pointer to an instance of the VkPipelineColorBlendStateCreateInfo structure, and is ignored if the pipeline has rasterization disabled or if the subpass of the render pass the pipeline is created against does not use any color attachments

The graphics driver tries to dereference the null pColorBlendState pointer and crashes.

The driver does not crash if I modify the vkCreateRenderPass call to set pCreateInfo->pSubpasses[0].colorAttachementCount = 0 for the VkRenderPass objects without color attachments.  This may be a viable application workaround until the driver is fixed.

**Comment by mikes-lunarg** [comment]
A little more digging and I confirmed that the driver used above predates a fix for KhronosGroup/Vulkan-LoaderAndValidationLayers#2543.

I confirmed the above crash does not occur with the latest NVIDIA linux driver 396.45 and DXVK 0.63.  It would be nice to see it working on windows as well, but I can't make it to the title screen with the latest driver (even without validation layers).

I'm going to close the issue. We can reopen or create a new issue if someone can reproduce the layer crash with a newer windows driver.
